File: plep_main/src/deltadak/database/DatabaseFacade.java
package deltadak.database;

import deltadak.Database;
import deltadak.HomeworkTask;
import deltadak.ui.AbstractController;
import deltadak.ui.Controller;
import javafx.concurrent.Task;
import javafx.scene.control.ProgressIndicator;

import java.time.LocalDate;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class DatabaseFacade {

    private AbstractController controller;

    public DatabaseFacade(AbstractController controller) {

        this.controller = controller;

    }

    /**
     * Updates database using the given homework tasks for a day.
     *
     * @param day
     *         Date from which the tasks are.
     * @param homeworkTasks
     *         Tasks to be put in the database.
     */
    public void updateDatabase(LocalDate day, List<List<HomeworkTask>> homeworkTasks) {
        // Database calls will be executed on a different thread.
        ExecutorService exec = Executors.newSingleThreadExecutor();

        if (controller != null) {
            ProgressIndicator progressIndicator = controller.getProgressIndicator();
            if (progressIndicator != null) {
                progressIndicator.setVisible(true);
                Task<List<HomeworkTask>> task = new Task<List<HomeworkTask>>() {
                    @Override
                    public List<HomeworkTask> call() throws Exception {
                        Database.INSTANCE.updateTasksDay(day, homeworkTasks);
                        // User feedback!
                        Thread.sleep(200);
                        return null;
                    }
                };
                task.setOnSucceeded(e -> progressIndicator.setVisible(false));
                exec.execute(task);
                exec.shutdown();
            }
        }
    }

}

File: plep_main/src/deltadak/database/DatabaseFacade.kt
package deltadak.database

import deltadak.Database
import deltadak.HomeworkTask
import javafx.concurrent.Task
import javafx.scene.control.ProgressIndicator
import java.time.LocalDate

/**
 * This class provides a facade to the database. General methods like pushing and pulling data from the database are included here, so extras can be added like proper user feedback and multithreading.
 */
class DatabaseFacade(
        /** For user feedback. */
        val progressIndicator: ProgressIndicator) {

    /**
     * Updates database using the given homework tasks for a day. Uses the progress indicator for user feedback.
     *
     * @param day Date from which the tasks are.
     * @param homeworkTasks Tasks to be put in the database.
     */
    fun pushData(day: LocalDate, homeworkTasks: List<List<HomeworkTask>> ) {

        val task: Task<Any> = object : Task<Any>() {
            @Throws(Exception::class)
            /** Specifies task. */
            public override fun call() {
                Database.INSTANCE.updateTasksDay(day, homeworkTasks)
            }
        }

        // Only switch it on and off if it's not yet on.
        if (!progressIndicator.isVisible) {

            // Switch on progress indicator.
            progressIndicator.isVisible = true

            // Switch off progress indicator.
            task.setOnSucceeded { progressIndicator.isVisible = false }

        }

        // Database calls will be executed on a different thread.
        executeMultithreading(task)

    }

    /**
     * Update only the parent tasks in the database.
     *
     * @param day The day which contains the tasks.
     * @param parentTasks The list with parents to update.
     */
    fun pushParentData(day: LocalDate, parentTasks: List<HomeworkTask>) {

        val task: Task<Any> = object : Task<Any>() {
            @Throws(Exception::class)
            /** Specifies task. */
            public override fun call() {
                Database.INSTANCE.updateParentsDay(day, parentTasks)
            }
        }

        // Only switch it on and off if it's not yet on.
        if (!progressIndicator.isVisible) {

            // Switch on progress indicator.
            progressIndicator.isVisible = true

            // Switch off progress indicator.
            task.setOnSucceeded { progressIndicator.isVisible = false }

        }

        // Database calls will be executed on a different thread.
        executeMultithreading(task)

    }

}